database/test_database.go

```diff
@@ -1258,7 +1258,74 @@ func FuzzNewIteratorWithPrefix(f *testing.F, db Database) {
 			require.Equal(expected[string(iter.Key())], val)
 			numIterElts++
 		}
-		require.Equal(len(expectedList), numIterElts)
+		require.Len(expectedList, numIterElts)
+
+		// Clear the database for the next fuzz iteration.
+		require.NoError(AtomicClear(db, db))
+	})
+}
+
+func FuzzNewIteratorWithStartAndPrefix(f *testing.F, db Database) {
+	const (
+		maxKeyLen   = 32
+		maxValueLen = 32
+	)
+
+	f.Fuzz(func(
+		t *testing.T,
+		randSeed int64,
+		start []byte,
+		prefix []byte,
+		numKeyValues uint,
+	) {
+		require := require.New(t)
+		r := rand.New(rand.NewSource(randSeed)) // #nosec G404
+
+		expected := map[string][]byte{}
+
+		// Put a bunch of key-values
+		for i := 0; i < int(numKeyValues); i++ {
+			key := make([]byte, r.Intn(maxKeyLen))
+			_, _ = r.Read(key) // #nosec G404
+
+			value := make([]byte, r.Intn(maxValueLen))
+			_, _ = r.Read(value) // #nosec G404
+
+			if len(value) == 0 {
+				// Consistently treat zero length values as nil
+				// so that we can compare [expected] and [got] with
+				// require.Equal, which treats nil and empty byte
+				// as being unequal, whereas the database treats
+				// them as being equal.
+				value = nil
+			}
+
+			if bytes.HasPrefix(key, prefix) && bytes.Compare(key, start) >= 0 {
```

Comment:
q: not super familiar with the fuzz framework. We have a random prefix generated by the framework and a key randomly generated here. Isn't pretty unlikely that our key has the "right" prefix? Have we checked the expected map is non-empty in a non neglegible number of cases? Otherwise the test may be not really do the testing of the positive case (key has prefix)

Reply:
I don't think the fuzz engine is completely naively generating inputs. According to [Go docs](https://go.dev/security/fuzz/), "Go fuzzing uses coverage guidance to intelligently walk through the code being fuzzed to find and report failures to the user."

If you add some logs you'll see that we do actually execute the body of the test.

Reply:
property testing ain't so smart. Cool!